t/unit/backends/test_couchdb.py

```diff
@@ -30,26 +30,77 @@ def test_init_no_pycouchdb(self):
             module.pycouchdb = prev
 
     def test_get_container_exists(self):
-            self.backend._connection = sentinel._connection
-            connection = self.backend.connection
-            assert connection is sentinel._connection
-            self.Server.assert_not_called()
+        self.backend._connection = sentinel._connection
+        connection = self.backend.connection
+        assert connection is sentinel._connection
+        self.Server.assert_not_called()
 
     def test_get(self):
         """test_get
 
         CouchBackend.get should return  and take two params
         db conn to couchdb is mocked.
-        TODO Should test on key not exists
-
         """
         x = CouchBackend(app=self.app)
         x._connection = Mock()
         get = x._connection.get = MagicMock()
-        # should return None
         assert x.get('1f3fab') == get.return_value['value']
         x._connection.get.assert_called_once_with('1f3fab')
 
+    def test_get_non_existent_key(self):
+        x = CouchBackend(app=self.app)
+        x._connection = Mock()
+        get = x._connection.get = MagicMock()
+        get.side_effect = pycouchdb.exceptions.NotFound
+        assert x.get('1f3fab') is None
+        x._connection.get.assert_called_once_with('1f3fab')
+
+    def test_set(self):
+        x = CouchBackend(app=self.app)
+        x._connection = Mock()
+
+        x.set('1f3fab', 'value')
+
+        x._connection.save.assert_called_once_with({'_id': '1f3fab',
+                                                    'value': 'value'})
+
+    def test_set_with_conflict(self):
+        x = CouchBackend(app=self.app)
+        x._connection = Mock()
+        x._connection.save.side_effect = (pycouchdb.exceptions.Conflict, None)
+        get = x._connection.get = MagicMock()
+
+        x.set('1f3fab', 'value')
+
+        x._connection.get.assert_called_once_with('1f3fab')
+        x._connection.get('1f3fab').__setitem__.assert_called_once_with(
+            'value', 'value')
+        x._connection.save.assert_called_with(get('1f3fab'))
+        assert x._connection.save.call_count == 2
+
+    def test_set_bytes_key(self):
+        x = CouchBackend(app=self.app)
+        x._connection = Mock()
+
+        x.set(b'1f3fab', 'value')
+
+        x._connection.save.assert_called_once_with({'_id': '1f3fab',
+                                                    'value': 'value'})
+
+    def test_set_bytes_key_with_conflict(self):
+        x = CouchBackend(app=self.app)
```

Comment:
These tests are identical, except for testing for the different types of key (which of course is the scope of this PR). Perhaps for the sake of DRY, we could just merge/group these cases using a loop, so instead of 4 we have 2? I think it will be less work for future readers and contributors. What do you think?

Reply:
We can parameterize the test using pytest. Good idea.